extensions/mqtt/AbstractMQTTProcessor.cpp

```diff
@@ -0,0 +1,158 @@
+  properties.insert(PassWord);
+  properties.insert(KeepLiveInterval);
+  properties.insert(ConnectionTimeOut);
+  properties.insert(QOS);
+  properties.insert(Topic);
+  setSupportedProperties(properties);
+  // Set the supported relationships
+  std::set<core::Relationship> relationships;
+  relationships.insert(Success);
+  relationships.insert(Failure);
+  setSupportedRelationships(relationships);
+}
+
+void AbstractMQTTProcessor::onSchedule(core::ProcessContext *context, core::ProcessSessionFactory *sessionFactory) {
+  std::string value;
+  int64_t valInt;
+  value = "";
+  if (context->getProperty(BrokerURL.getName(), value) && !value.empty()) {
+    uri_ = value;
+    logger_->log_info("AbstractMQTTProcessor: BrokerURL [%s]", uri_);
+  }
+  value = "";
+  if (context->getProperty(ClientID.getName(), value) && !value.empty()) {
+    clientID_ = value;
+    logger_->log_info("AbstractMQTTProcessor: ClientID [%s]", clientID_);
+  }
+  value = "";
+  if (context->getProperty(Topic.getName(), value) && !value.empty()) {
+    topic_ = value;
+    logger_->log_info("AbstractMQTTProcessor: Topic [%s]", topic_);
+  }
+  value = "";
+  if (context->getProperty(UserName.getName(), value) && !value.empty()) {
+    userName_ = value;
+    logger_->log_info("AbstractMQTTProcessor: UserName [%s]", userName_);
+  }
+  value = "";
+  if (context->getProperty(PassWord.getName(), value) && !value.empty()) {
+    passWord_ = value;
+    logger_->log_info("AbstractMQTTProcessor: PassWord [%s]", passWord_);
+  }
+  value = "";
+  if (context->getProperty(CleanSession.getName(), value) && !value.empty() &&
+      org::apache::nifi::minifi::utils::StringUtils::StringToBool(value, cleanSession_)) {
+    logger_->log_info("AbstractMQTTProcessor: CleanSession [%d]", cleanSession_);
+  }
+  value = "";
+  if (context->getProperty(KeepLiveInterval.getName(), value) && !value.empty()) {
+    core::TimeUnit unit;
+    if (core::Property::StringToTime(value, valInt, unit) && core::Property::ConvertTimeUnitToMS(valInt, unit, valInt)) {
+      keepAliveInterval_ = valInt/1000;
+      logger_->log_info("AbstractMQTTProcessor: KeepLiveInterval [%d]", keepAliveInterval_);
+    }
+  }
+  value = "";
+  if (context->getProperty(ConnectionTimeOut.getName(), value) && !value.empty()) {
+    core::TimeUnit unit;
+    if (core::Property::StringToTime(value, valInt, unit) && core::Property::ConvertTimeUnitToMS(valInt, unit, valInt)) {
+      connectionTimeOut_ = valInt/1000;
+      logger_->log_info("AbstractMQTTProcessor: ConnectionTimeOut [%d]", connectionTimeOut_);
+    }
+  }
+  value = "";
+  if (context->getProperty(QOS.getName(), value) && !value.empty() && (value == MQTT_QOS_0 || value == MQTT_QOS_1 || MQTT_QOS_2) &&
+      core::Property::StringToInt(value, valInt)) {
+    qos_ = valInt;
+    logger_->log_info("PublishKafka: QOS [%d]", qos_);
+  }
+  if (!client_) {
+    MQTTClient_create(&client_, uri_.c_str(), clientID_.c_str(), MQTTCLIENT_PERSISTENCE_NONE, NULL);
```

Comment:
what's the thread safety of this function call?

Reply:
it is thread safeint 
MQTTClient_create(MQTTClient* handle, const char* serverURI, const char* clientId,
		int persistence_type, void* persistence_context)
{
	int rc = 0;
	MQTTClients *m = NULL;

	FUNC_ENTRY;
	rc = Thread_lock_mutex(mqttclient_mutex);